File: iosApp/iosApp/Modules/DegenCult/CultProposals/CultProposalsWireframe.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultCultProposalsWireframe {
    private weak var parent: UIViewController?
    private let cultProposalWireframeFactory: CultProposalWireframeFactory
    private let confirmationWireframeFactory: ConfirmationWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    private let currencySwapWireframeFactory: CurrencySwapWireframeFactory
    private let cultService: CultService
    private let walletService: WalletService
    private let networksService: NetworksService

    private weak var vc: UIViewController?

    init(
        _ parent: UIViewController?,
        cultProposalWireframeFactory: CultProposalWireframeFactory,
        confirmationWireframeFactory: ConfirmationWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory,
        currencySwapWireframeFactory: CurrencySwapWireframeFactory,
        cultService: CultService,
        walletService: WalletService,
        networksService: NetworksService
    ) {
        self.parent = parent
        self.cultProposalWireframeFactory = cultProposalWireframeFactory
        self.confirmationWireframeFactory = confirmationWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
        self.currencySwapWireframeFactory = currencySwapWireframeFactory
        self.cultService = cultService
        self.walletService = walletService
        self.networksService = networksService
    }
}

extension DefaultCultProposalsWireframe {

    func present() {
        let vc = wireUp()
        parent?.show(vc, sender: self)
    }

    func navigate(to destination: CultProposalsWireframeDestination) {
        if let input = destination as? CultProposalsWireframeDestination.Proposal {
            let context = CultProposalWireframeContext(proposal: input.proposal, proposals: input.proposals)
            cultProposalWireframeFactory.make(vc, context: context).present()
        }
        if let input = destination as? CultProposalsWireframeDestination.CastVote {
            let networkFee = networksService.defaultNetworkFee(network: Network.ethereum())
            let context = ConfirmationWireframeContext.CultCastVote(
                data: .init(cultProposal: input.proposal, approve: input.approve, networkFee: networkFee)
            )
            confirmationWireframeFactory.make(vc, context: context).present()
        }
        if let input = destination as? CultProposalsWireframeDestination.Alert {
            alertWireframeFactory.make(vc, context: input.context).present()
        }
        if destination is CultProposalsWireframeDestination.GetCult {
            let context = CurrencySwapWireframeContext(
                network: .ethereum(), currencyFrom: nil, currencyTo: Currency.Companion().cult()
            )
            currencySwapWireframeFactory.make(vc, context: context).present()
        }
    }
}

private extension DefaultCultProposalsWireframe {

    func wireUp() -> UIViewController {
        let vc: CultProposalsViewController = UIStoryboard(.cultProposals).instantiate()
        let interactor = DefaultCultProposalsInteractor(
            cultService: cultService,
            walletService: walletService
        )
        let presenter = DefaultCultProposalsPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor
        )
        vc.presenter = presenter
        self.vc = vc
        return vc
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/cultproposals/CultProposalsWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.cultproposals

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.types.Network
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.NavigationFragment
import com.sonsofcrypto.web3wallet.android.common.extensions.navigationFragment
import com.sonsofcrypto.web3wallet.android.common.ui.navigationFragment
import com.sonsofcrypto.web3wallet.android.modules.compose.confirmation.ConfirmationWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.cultproposal.CultProposalWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext.CultCastVoteContext
import com.sonsofcrypto.web3walletcore.modules.degenCultProposal.CultProposalWireframeContext
import com.sonsofcrypto.web3walletcore.modules.degenCultProposals.CultProposalsWireframe
import com.sonsofcrypto.web3walletcore.modules.degenCultProposals.CultProposalsWireframeDestination
import com.sonsofcrypto.web3walletcore.modules.degenCultProposals.DefaultCultProposalsInteractor
import com.sonsofcrypto.web3walletcore.modules.degenCultProposals.DefaultCultProposalsPresenter
import com.sonsofcrypto.web3walletcore.services.cult.CultService

class DefaultCultProposalsWireframe(
    private val parent: WeakRef<Fragment>?,
    private val cultService: CultService,
    private val walletService: WalletService,
    private val networksService: NetworksService,
    private val cultProposalWireframeFactory: CultProposalWireframeFactory,
    private val confirmationWireframeFactory: ConfirmationWireframeFactory,
): CultProposalsWireframe {

    override fun present() {
        val fragment = wireUp()
        parent?.navigationFragment?.present(fragment, animated = true)
    }

    override fun navigate(destination: CultProposalsWireframeDestination) {
        when (destination) {
            is CultProposalsWireframeDestination.Proposal -> {
                val context = CultProposalWireframeContext(
                    destination.proposal, destination.proposals
                )
                cultProposalWireframeFactory.make(parent?.get(), context).present()
            }
            is CultProposalsWireframeDestination.CastVote -> {
                val networkFee = networksService.defaultNetworkFee(Network.ethereum())
                val context = ConfirmationWireframeContext.CultCastVote(
                    CultCastVoteContext(
                        cultProposal = destination.proposal,
                        approve = destination.approve,
                        networkFee = networkFee,
                    )
                )
                confirmationWireframeFactory.make(parent?.get(), context).present()
            }
            is CultProposalsWireframeDestination.Dismiss -> {
                parent?.navigationFragment?.popOrDismiss()
            }
            else -> { println("handle event $destination")}
        }
    }

    private fun wireUp(): Fragment {
        val view = CultProposalsFragment()
        val interactor = DefaultCultProposalsInteractor(
            cultService, walletService
        )
        val presenter = DefaultCultProposalsPresenter(
            WeakRef(view),
            this,
            interactor
        )
        view.presenter = presenter
        return view
    }
}